src/pcnt.rs

```diff
@@ -1,656 +1,1121 @@
+
+        interrupt::with_isr_yield_signal(move || (user_data.on_reach)(event_data))
+    }
+}
+
+impl<'d, S: IsEnabled> PcntUnitDriver<'d, S> {
+    /// Disable the PCNT unit.
     ///
-    /// @param callback Interrupt handler function.
+    /// # Note
     ///
-    /// returns
-    /// - ()
-    /// - EspError
-    #[cfg(feature = "alloc")]
-    pub unsafe fn subscribe_nonstatic<F>(&self, callback: F) -> Result<(), EspError>
-    where
-        F: FnMut(u32) + Send + 'd,
-    {
-        self.internal_subscribe(callback)
+    /// Disable a PCNT unit doesn't mean to stop it. See also [`Self::stop`] for how to stop the PCNT counter.
+    pub fn disable(mut self) -> Result<PcntUnitDriver<'d, Disabled>, EspError> {
+        self.internal_disable()?;
+
+        Ok(self.change_state())
     }
 
-    #[cfg(feature = "alloc")]
-    fn internal_subscribe<F>(&self, callback: F) -> Result<(), EspError>
-    where
-        F: FnMut(u32) + Send + 'd,
-    {
-        enable_isr_service()?;
+    /// Disables the PCNT unit, runs the given closure, and then enables the PCNT unit again.
+    pub fn disable_for<T>(
+        &mut self,
+        f: impl FnOnce(&mut PcntUnitDriver<'d, Disabled>) -> Result<T, EspError>,
+    ) -> Result<T, EspError> {
+        self.internal_disable()?;
 
-        self.unsubscribe()?;
-        let callback: alloc::boxed::Box<dyn FnMut(u32) + 'd> = alloc::boxed::Box::new(callback);
+        let disabled_driver = self.change_state_mut();
 
-        unsafe {
-            ISR_HANDLERS[self.unit as usize] = Some(core::mem::transmute::<
-                alloc::boxed::Box<dyn FnMut(u32)>,
-                alloc::boxed::Box<dyn FnMut(u32)>,
-            >(callback));
-        }
-        esp!(unsafe {
-            pcnt_isr_handler_add(
-                self.unit,
-                Some(Self::handle_isr),
-                self.unit as *mut core::ffi::c_void,
-            )
-        })?;
+        let result = f(disabled_driver);
 
-        Ok(())
+        self.internal_enable()?;
+
+        result
     }
 
-    /// Remove ISR handler for specified unit.
+    /// Start the PCNT unit, the counter will start to count according to the edge and/or level input signals.
     ///
-    /// returns
-    /// - ()
-    /// - EspError
-    #[cfg(feature = "alloc")]
-    pub fn unsubscribe(&self) -> Result<(), EspError> {
-        unsafe {
-            esp!(pcnt_isr_handler_remove(self.unit))?;
-            ISR_HANDLERS[self.unit as usize] = None;
-        }
-        Ok(())
+    /// # ISR Safety
+    ///
+    /// This function is safe to be called from an ISR context.
+    ///
+    /// This function will be placed into IRAM if `CONFIG_PCNT_CTRL_FUNC_IN_IRAM`
+    /// is on, so that it's allowed to be executed when cache is disabled.
+    ///
+    /// # Errors
+    ///
+    /// - `ESP_FAIL`: Start PCNT unit failed because of other error
+    #[cfg_attr(esp_idf_pcnt_ctrl_func_in_iram, link_section = ".iram1.rpcnt_start")]
```

Comment:
One thing we need to double check here. It may be that for different esp's the iram section might be different. Where did you pick up the .iram1.rpcnt_start? I think duo to the classical esp32 nature it has a split region. E.g iram0 and iram1, but that is not the case for other esp's. 

So if we activate that we need to double check that its the correct place for that esp variant.

Reply:
I looked at how this library does it: https://github.com/search?q=repo%3Aesp-rs%2Fesp-idf-hal%20.iram&type=code

And assumed that I should do the same?

When looking at these, I think `#[inline(always)]` is not right...it should be `#[inline(never)]`

I think this project would greatly benefit from a proc-macro that does this like esp-hal has with their `#[ram]` macro

Reply:
Implementing the proc-macro is only a few hours (at most) of work after the API is defined. I will open an issue for it (if there isn't one yet)

Reply:
Since its a minor thing just comment out / remove the annotation to not block the PR itself, so me move forward here. That than can be adjusted later.